File: btrader/core/TradingPair.py
__all__ = [
  'TradingPair'
]

class TradingPair ():

  def __init__ (self, symbol, base=None, quote=None):
    if not symbol:
      self.__symbol = base+quote
      self.__baseAsset = base
      self.__quoteAsset = quote
    else:
      self.__symbol = symbol['symbol']
      self.__baseAsset = symbol['baseAsset']
      self.__baseAssetPrecision = symbol['baseAssetPrecision']
      self.__quoteAsset = symbol['quoteAsset']
      self.__quoteAssetPrecision = symbol['quoteAssetPrecision']
      self.__step = None
      for f in symbol['filters']:
        if f['filterType'] == 'LOT_SIZE':
          self.__step = float(f['stepSize'])
      if self.__step == None:
        raise ValueError ("Couldn't get step size for symbol {}".format(self.__symbol))

  @property
  def symbol (self):
    return self.__symbol

  @property
  def step (self):
    return self.__step

  @property
  def baseAsset (self):
    return self.__baseAsset

  @property
  def baseAssetPrecision (self):
    return self.__baseAssetPrecision

  @property
  def quoteAsset (self):
    return self.__quoteAsset

  @property
  def quoteAssetPrecision (self):
    return self.__quoteAssetPrecision

  def hasAsset (self, asset):
    if self.quoteAsset == asset or self.baseAsset == asset:
      return True
    return False

  def getTheOther (self, asset):
    if self.quoteAsset == asset:
      return self.baseAsset
    elif self.baseAsset == asset:
      return self.quoteAsset
    else:
      return None

  def __repr__ (self):
    return ("<TradingPair {}/{}>".format(self.baseAsset, self.quoteAsset))

  def __str__ (self):
    return self.__repr__()

  @property
  def text (self):
    return "{}/{}".format(self.baseAsset, self.quoteAsset)

  def __eq__ (self, tp):
    if ((tp.quoteAsset == self.quoteAsset) and (tp.baseAsset == self.baseAsset)) or ((tp.quoteAsset == self.baseAsset) and (tp.baseAsset == self.quoteAsset)):
      return True
    return False

File: src/trading_pair.rs
use binance::model::*;
use std::fmt;
use std::time::SystemTime;
/*
 *  TradingPair
 */
#[derive(Debug, Clone)]
pub struct TradingPair {
  symbol: String,
  base_asset: String,
  quote_asset: String,
  step: f64,
  initialized: bool,
  asks: Vec<Asks>,
  bids: Vec<Bids>,
  timestamp: SystemTime,
}

impl fmt::Display for TradingPair {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "<TradingPair {}/{}>", self.base_asset, self.quote_asset)
  }
}

impl TradingPair {
  // Constructor
  pub fn new(symbol: String, base_asset: String, quote_asset: String, step: f64) -> TradingPair {
    TradingPair {
      symbol,
      base_asset,
      quote_asset,
      step,
      initialized: false,
      asks: Vec::new(),
      bids: Vec::new(),
      timestamp: SystemTime::now(),
    }
  }
  // Setters
  pub fn update(&mut self, timestamp: SystemTime, asks: Vec<Asks>, bids: Vec<Bids>) {
    self.asks = asks;
    self.bids = bids;
    self.timestamp = timestamp;
    self.initialized = true;
  }
  // Getters
  pub fn get_symbol(&self) -> String {
    self.symbol.to_string()
  }
  pub fn get_step(&self) -> f64 {
    self.step
  }
  pub fn get_base_asset(&self) -> String {
    self.base_asset.to_string()
  }
  pub fn get_quote_asset(&self) -> String {
    self.quote_asset.to_string()
  }
  // Utilities
  pub fn has_asset(&self, asset: String) -> bool {
    if (asset == self.quote_asset) || (asset == self.base_asset) {
      return true;
    }
    false
  }
  pub fn get_the_other(&self, asset: String) -> String {
    if asset == self.quote_asset {
      self.base_asset.to_string()
    } else if asset == self.base_asset {
      self.quote_asset.to_string()
    } else {
      "".to_string()
    }
  }

  pub fn text(&self) -> String {
    format!("{}/{}", self.base_asset, self.quote_asset)
  }
}

// Eq overloading
impl PartialEq for TradingPair {
  fn eq(&self, other: &Self) -> bool {
    if ((other.quote_asset == self.quote_asset) && (other.base_asset == self.base_asset))
      || ((other.quote_asset == self.base_asset) && (other.base_asset == self.quote_asset))
    {
      return true;
    }
    false
  }
}
impl Eq for TradingPair {}
